drivers: Rework Silabs Gecko UART Driver

diegosueiro opened the pull request:
This patch series introduces a new config HAS_DTS_SERIAL and reworks the Silabs Gecko UART driver and associated soc and board files to enable using values defined by the device tree configuration.

diegosueiro committed:
ext: Silabs Gecko use a different config for uart

Use CONFIG_UART_GECKO instead of CONFIG_SERIAL_HAS_DRIVER to compile
the low level uart driver.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro commented on drivers/serial/Kconfig.gecko:
```diff
@@ -26,7 +26,7 @@ if UART_GECKO_0
 	int "Pin Locations"
 	range 0 3
-	depends on UART_GECKO
+	depends on UART_GECKO && !HAS_DTS_SERIAL
```
@chrta,

I totally agree with you. 

@galak, 

What do you think? Can we remove the *_LOC from Kconfig?

diegosueiro committed:
drivers: Add more uart instances for Silabs Devices

Add more uart/usart instances for Silabs Gecko Devices and remove
the *_GPIO_LOC configs.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro commented:
@galak and @MaureenHelm ,

Any feedbacks?

diegosueiro commented:
@galak and @MaureenHelm ,

Any feedbacks?

diegosueiro commented:
@galak, can we proceed with the review?

galak commented:
These changes seem part of the PR #9042.  I made comments over there.

galak commented on dts/bindings/serial/silabs,gecko-uart.yaml:
```diff
@@ -1,17 +1,17 @@
-title: EFM32 UART
-id: silabs,efm32-uart
+title: GECKO UART
+id: silabs,gecko-uart
```
Trying to remove id: so lets leave that alone

galak commented on dts/bindings/serial/silabs,gecko-usart.yaml:
```diff
@@ -1,17 +1,17 @@
-title: EFM32 USART
-id: silabs,efm32-usart
+title: GECKO USART
+id: silabs,gecko-usart
```
same comment about id.

galak commented on soc/arm/silabs_exx32/efr32fg1p/soc_pinmap.h:
```diff
@@ -33,15 +33,15 @@
 
 #ifdef CONFIG_USART_GECKO_1
-#if (CONFIG_USART_GECKO_1_GPIO_LOC == 0)
+#if (USART_1_LOCATION == 0)
```
leave these as CONFIG_

galak commented on soc/arm/silabs_exx32/efr32fg1p/soc_pinmap.h:
```diff
@@ -18,13 +18,13 @@
 #ifdef CONFIG_SOC_PART_NUMBER_EFR32FG1P133F256GM48
 #ifdef CONFIG_USART_GECKO_0
-#if (CONFIG_USART_GECKO_0_GPIO_LOC == 0)
+#if (USART_0_LOCATION == 0)
```
leave as CONFIG_

galak commented on soc/arm/silabs_exx32/efr32fg1p/dts.fixup:
```diff
@@ -7,9 +7,5 @@
 /* SoC level DTS fixup file */
 
 #define CONFIG_NUM_IRQ_PRIO_BITS	ARM_V7M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
-#define CONFIG_USART_GECKO_0_NAME	SILABS_EFM32_USART_40010000_LABEL
```
Leave this mapping

galak commented on soc/arm/silabs_exx32/efm32wg/soc_pinmap.h:
```diff
@@ -18,13 +18,13 @@
 #ifdef CONFIG_SOC_PART_NUMBER_EFM32WG990F256
 #ifdef CONFIG_UART_GECKO_0
-#if (CONFIG_UART_GECKO_0_GPIO_LOC == 0)
+#if (UART_0_LOCATION == 0)
```
keep these as CONFIG_

galak reviewed (changes requested):
Use CONFIG_ for dt generated defines.  This makes it easier for us to know what is being generated by DT rather than dropping the CONFIG_ prefix.  Eventually this will go away when we do more codegen and remove dts fixups.

Also, might need updating now that we merged the EFM32HG support

ekarlso commented:
Hi, can we rebase this ontop https://github.com/ekarlso/zephyr/tree/silabs-uart @diegosueiro ?

carlescufi commented:
@ekarlso this was rebased on top of your tree? do you have a PR outstanding or is this the only Gecko UART driver PR open?

diegosueiro committed:
drivers: serial: Rework Silabs Gecko UART Driver

Introduces the location property and adds the ability to use values
generated by the device tree configuration.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak reviewed (commented):
I've made some updates, dropped board.h from driver.  Used standard CONFIG_ naming convention for now, dropped 'id:' from yaml.

Can you test this out to make sure it still works.

diegosueiro commented:
@galak,

> I've made some updates, dropped board.h from driver. Used standard CONFIG_ naming convention for now, dropped 'id:' from yaml.
> 
> Can you test this out to make sure it still works.

The only hardware that I have is the thunderboard from the other PR #9042.

But the changes looks fine to me.

diegosueiro commented on drivers/serial/uart_gecko.c:
```diff
@@ -10,6 +10,7 @@
 #include <em_gpio.h>
 #include <em_cmu.h>
 #include <soc.h>
+#include <board.h>
```
Because for efr32mg_sltb004a board the board.h includes the "soc_pinmap.h" which holds the UART pin settings.

The other silabs boards are setting the pins in the "soc_pinmap.h" header which is included by the "soc.h" header, and I think is not the best approach since pin configuration/muxing should be set at board level, not soc. What do you think?

chrta replied:
The mapping which location corresponds to which pin is a property of the soc. But the location that is actually chosen is part of the board. Or am I missing something here?

galak replied:
Its a 50/50 case, in that we set some standard pin cfg in the soc headers for a number of SoCs.  The board ends up select a default or pre-cfg option in a lot of cases.  Eventually we will move pin cfg info into the device tree.  For now I'd keep with the same pattern that other SiLabs boards/SoC are using.

diegosueiro replied:
Ok.

I'll update the PR #9042 as soon as this one gets merged.

galak commented:
> But the changes looks fine to me apart from dropping the board.h as I explained [here](https://github.com/zephyrproject-rtos/zephyr/pull/8930#discussion_r225683292)

ok, let me look at that.

chrta reviewed (approved):
This is working fine on the EFM32WG-STK3800
